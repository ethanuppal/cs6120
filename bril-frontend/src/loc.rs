// Copyright (C) 2024 Ethan Uppal.
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, version 3 of the License only.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program.  If not, see <https://www.gnu.org/licenses/>.

use std::{
    fmt,
    ops::{Deref, DerefMut},
};

pub use logos::Span;

#[derive(Debug, Clone)]
pub struct Loc<T> {
    pub inner: T,
    pub span: Span,
}

impl<T> Loc<T> {
    pub fn new(inner: T, span: Span) -> Self {
        Self { inner, span }
    }

    pub fn without_inner(&self) -> Loc<()> {
        Loc::new((), self.span.clone())
    }

    pub fn map<U, F: FnOnce(T) -> U>(self, f: F) -> Loc<U> {
        Loc::new(f(self.inner), self.span)
    }
}

impl<T> Deref for Loc<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<T> DerefMut for Loc<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl<T: fmt::Display> fmt::Display for Loc<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.inner.fmt(f)
    }
}

pub trait Spanned {
    fn span(&self) -> Span;
}

impl Spanned for Span {
    fn span(&self) -> Span {
        self.clone()
    }
}

impl Spanned for &Span {
    fn span(&self) -> Span {
        (*self).clone()
    }
}

impl<T> Spanned for Loc<T> {
    fn span(&self) -> Span {
        self.span.clone()
    }
}

impl<T> Spanned for &Loc<T> {
    fn span(&self) -> Span {
        self.span.clone()
    }
}

pub trait WithLocation {
    fn at(self, spanned: impl Spanned) -> Loc<Self>
    where
        Self: Sized,
    {
        Loc::new(self, spanned.span())
    }

    fn between(self, start: impl Spanned, end: impl Spanned) -> Loc<Self>
    where
        Self: Sized,
    {
        Loc::new(self, start.span().start..end.span().end)
    }
}

impl<T> WithLocation for T {}
