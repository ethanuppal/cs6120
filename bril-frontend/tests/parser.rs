macro_rules! parser_snapshot {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            use bril_frontend::loc::Loc;
            use insta::assert_snapshot;
            use logos::Logos;

            let code = $code;
            let mut lexer = bril_frontend::lexer::Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    panic!("Failed to lex. Leftover: {}", lexer.remainder());
                }
            }

            let mut parser = bril_frontend::parser::Parser::new(&tokens);

            let Ok(program) = parser.parse_program() else {
                for diagnostic in parser.diagnostics() {
                    println!("{}:", diagnostic.message);
                    for (text, span) in &diagnostic.labels {
                        println!("Label: {}", text);
                        println!(
                            "Code: `{}`",
                            &code[span
                                .clone()
                                .unwrap_or(diagnostic.span.clone())]
                        );
                    }
                }
                panic!("Failed to parse program");
            };

            let mut buffer = String::new();
            bril_frontend::printer::Printer::new(&mut buffer, 2)
                .print_program(&program)
                .expect("Failed to print program");

            assert_snapshot!(format!(
                "ORIGINAL\n--------\n{}\n\nPRINTED\n-------\n{}",
                code, buffer
            ));
        }
    };
}

parser_snapshot! {
    add_no_print_bril_parses,
    include_str!("../bril-programs/add_no_print.bril")
}

parser_snapshot! {
    add_bril_parses,
    include_str!("../bril-programs/add.bril")
}

parser_snapshot! {
    import_bril_parses,
    include_str!("../bril-programs/import.bril")
}

parser_snapshot! {
    simple_bril_parses,
    include_str!("../bril-programs/simple.bril")
}
