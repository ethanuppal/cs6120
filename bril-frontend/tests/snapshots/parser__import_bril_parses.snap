---
source: bril-frontend/tests/parser.rs
expression: "format! (\"ORIGINAL\\n--------\\n{}\\n\\nPRINTED\\n-------\\n{}\", code, buffer)"
---
ORIGINAL
--------
from "foo" import @bar, @baz as @bop;
from "foo" import @bar, @baz;


PRINTED
-------
from "foo" import @bar, @baz as @bop;
from "foo" import @bar, @baz;
